src/libbson/src/bson/bson.c

```diff
@@ -60,6 +61,36 @@ typedef struct {
  */
 static const uint8_t gZero;
 
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _bson_round_up_alloc_size --
+ *
+ *       Given a potential allocation length in bytes, round up to the
+ *       next power of two without exceeding BSON_MAX_SIZE.
+ *
+ * Returns:
+ *       If the input is <= BSON_MAX_SIZE, returns a value >= the input
+ *       and still <= BSON_MAX_SIZE. If the input was greater than
+ *       BSON_MAX_SIZE, it is returned unmodified.
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+static BSON_INLINE size_t
+_bson_round_up_alloc_size (size_t size)
+{
+   if (size <= BSON_MAX_SIZE) {
+      size_t power_of_two = bson_next_power_of_two (size);
+      return BSON_MIN (power_of_two, BSON_MAX_SIZE);
+   } else {
+      return size;
```

Comment:
Allowing `size > BSON_MAX_SIZE` to be returned unmodified feels like an odd special-case for this function. Can this be made a precondition for this function and the `size > BSON_MAX_SIZE` handling be moved into the call sites instead (where other bounds checks are already being performed)?

Reply:
The call to `bson_next_power_of_two` in `bson_copy_to` is suspect given `bson_t` creation functions should be enforcing `len <= BSON_MAX_SIZE` via checks or assertions as an invariant of `bson_t` objects (+ why should a "copy" reserve more space than the value being copied from?), but I suppose this does not protect against manually-created `bson_t` objects which may have a buffer length greater than `BSON_MAX_SIZE`. Given this is falls outside the scope of this PR, I am okay with deferring this item.